plugins/Toolbox/src/Toolbox.py

```diff
@@ -0,0 +1,609 @@
+                self._temp_plugin_file.write(self._download_reply.readAll())
+                self._temp_plugin_file.close()
+                self._onDownloadComplete(file_path)
+                return
+
+    def _onDownloadComplete(self, file_path):
+        Logger.log("i", "Toolbox: Download complete.")
+        try:
+            package_info = self._package_manager.getPackageInfo(file_path)
+        except:
+            Logger.logException("w", "Toolbox: Package file [%s] was not a valid CuraPackage.", file_path)
+            return
+
+        license_content = self._package_manager.getPackageLicense(file_path)
+        if license_content is not None:
+            self.openLicenseDialog(package_info["package_id"], license_content, file_path)
+            return
+
+        self.install(file_path)
+        return
+
+
+    # Getter & Setters for Properties:
+    # --------------------------------------------------------------------------
+    def setDownloadProgress(self, progress):
+        if progress != self._download_progress:
+            self._download_progress = progress
+            self.onDownloadProgressChanged.emit()
+    @pyqtProperty(int, fset = setDownloadProgress, notify = onDownloadProgressChanged)
+    def downloadProgress(self):
+        return self._download_progress
+
+    def setIsDownloading(self, is_downloading):
+        if self._is_downloading != is_downloading:
+            self._is_downloading = is_downloading
+            self.onIsDownloadingChanged.emit()
+    @pyqtProperty(bool, fset = setIsDownloading, notify = onIsDownloadingChanged)
+    def isDownloading(self):
+        return self._is_downloading
+
+    def setActivePackage(self, package):
+        self._active_package = package
+        self.activePackageChanged.emit()
+    @pyqtProperty(QObject, fset = setActivePackage, notify = activePackageChanged)
+    def activePackage(self):
+        return self._active_package
+
+    def setViewCategory(self, category = "plugins"):
+        self._view_category = category
+        self.viewChanged.emit()
+    @pyqtProperty(str, fset = setViewCategory, notify = viewChanged)
+    def viewCategory(self):
+        return self._view_category
+
+    def setViewPage(self, page = "overview"):
+        self._view_page = page
+        self.viewChanged.emit()
+    @pyqtProperty(str, fset = setViewPage, notify = viewChanged)
+    def viewPage(self):
+        return self._view_page
+
+    def setViewSelection(self, selection = ""):
+        self._view_selection = selection
+        self.viewChanged.emit()
+    @pyqtProperty(str, fset = setViewSelection, notify = viewChanged)
+    def viewSelection(self):
+        return self._view_selection
+
+
+
+    # Expose Models:
+    # --------------------------------------------------------------------------
+    # TODO: Maybe replace this with simply exposing self._models to Qt and then
+    # setting model: toolbox.models.foobar instead of toolbox.foobarModel
```

Comment:
Actually I think it is fine like this. Just my opinion though. This way you don't need a separate Models object.